gnucash/gnome-utils/dialog-assoc-utils.c

```diff
@@ -0,0 +1,454 @@
+        else
+            file_path = gnc_file_path_absolute (NULL, uri);
+
+        if (return_uri)
+            ret_value = gnc_uri_create_uri ("file", NULL, 0, NULL, NULL, file_path);
+        else
+            ret_value = g_strdup (file_path);
+    }
+
+    if (g_strcmp0 (uri_scheme, "file") == 0) // absolute path
+    {
+        if (return_uri)
+            ret_value = g_strdup (uri);
+        else
+            ret_value = gnc_uri_get_path (uri);
+    }
+    g_free (file_path);
+    return ret_value;
+}
+
+gchar *
+gnc_assoc_get_unescape_uri (const gchar *path_head, const gchar *uri, gchar *uri_scheme)
+{
+    gchar *display_str = NULL;
+    gchar *file_path = NULL;
+
+    if (uri && *uri != '\0')
+    {
+        // if scheme is null or 'file' we should get a file path
+        file_path = convert_uri_to_abs_path (path_head, uri, uri_scheme, FALSE);
+
+        if (file_path)
+            display_str = g_uri_unescape_string (file_path, NULL);
+        else
+            display_str = g_uri_unescape_string (uri, NULL);
+
+#ifdef G_OS_WIN32 // make path look like a traditional windows path
+        display_str = g_strdelimit (display_str, "/", '\\');
+#endif
+    }
+    DEBUG("Return display string is '%s'", display_str);
+    g_free (file_path);
+    return display_str;
+}
+
+gchar *
+gnc_assoc_get_use_uri (const gchar *path_head, const gchar *uri, gchar *uri_scheme)
+{
+    gchar *use_str = NULL;
+    gchar *file_path = NULL;
+
+    if (uri && *uri != '\0')
+    {
+        // if scheme is null or 'file' we should get a file path
+        file_path = convert_uri_to_abs_path (path_head, uri, uri_scheme, TRUE);
+
+        if (file_path)
+            use_str = g_strdup (file_path);
+        else
+            use_str = g_strdup (uri);
+    }
+    DEBUG("Return use string is '%s'", use_str);
+    g_free (file_path);
+    return use_str;
+}
+
+gchar *
+gnc_assoc_get_unescaped_just_uri (const gchar *uri)
+{
+    gchar *path_head = gnc_prefs_get_string (GNC_PREFS_GROUP_GENERAL, "assoc-head");
+    gchar *uri_scheme = gnc_uri_get_scheme (uri);
+    gchar *ret_uri = gnc_assoc_get_unescape_uri (path_head, uri, uri_scheme);
+
+    g_free (path_head);
+    g_free (uri_scheme);
+    return ret_uri;
+}
+
+gchar *
+gnc_assoc_convert_trans_associate_uri (gpointer trans, gboolean book_ro)
+{
+    const gchar *uri = xaccTransGetAssociation (trans); // get the existing uri
+    const gchar *part = NULL;
+
+    if (!uri)
+        return NULL;
+
+    if (g_str_has_prefix (uri, "file:") && !g_str_has_prefix (uri,"file://"))
+    {
+        // fix an earlier error when storing relative paths before version 3.5
+        // relative paths are stored without a leading "/" and in native form
+        if (g_str_has_prefix (uri,"file:/") && !g_str_has_prefix (uri,"file://"))
+            part = uri + strlen ("file:/");
+        else if (g_str_has_prefix (uri,"file:") && !g_str_has_prefix (uri,"file://"))
+            part = uri + strlen ("file:");
```

Comment:
IIRC I wrote the original gnc_uri_utils under the false assumption uri's would start with file:// instead of file:///. So current code other than the file associations also work under this assumption. I'm fine with cleaning that up, though that's probably best done in a separate PR.

Reply:
I have changed this, the test is for 'file:' and 'file:/' with that part being removed from the string so that relative paths are stored without a scheme start and without a '/' if relative. Absolute uri paths do store with 'file:///'